staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var ignoreTimestampEqualities = func() conversion.Equalities {
+	var eqs = equality.Semantic.Copy()
+
+	err := eqs.AddFunc(
+		func(a, b metav1.ManagedFieldsEntry) bool {
+			// Two objects' managed fields are equivalent if, ignoring timestamp,
+			//	the objects are deeply equal.
+			a.Time = nil
+			b.Time = nil
+			return reflect.DeepEqual(a, b)
+		},
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	return eqs
+}()
+
+// IgnoreManagedFieldsTimestampsTransformer reverts timestamp updates
+// if the non-managed parts of the object are equivalent
+func IgnoreManagedFieldsTimestampsTransformer(
+	_ context.Context,
+	newObj runtime.Object,
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if ignoreTimestampEqualities.DeepEqualWithNilDifferentFromEmpty(newObj, oldObj) {
```

Comment:
Do you need to compare the entire objects? or is it OK to just compare the managed fields?

Reply:
the bug is that managedFields changed timestamps when nothing else changed in the rest of the object

Reply:
I had thought about an initial fast(er)/cheap(er) check to see if managedFields was identical except for timestamps, ahead of the full object comparison, but:
* that will likely be true in the case of high-volume controller-driven status updates touching the same fields, which is where most of our traffic comes from
* getting managedFields out via an accessor adds its own allocations for custom resources: https://github.com/kubernetes/kubernetes/blob/64c5ed380441b4015301cf64a44a698400b03d8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L459-L479

Reply:
OK, you've read the first thought out of my mind, but I have more:

Has generation been set at this point? If so, you can short-circuit a spec check by looking at that.

We can compare everything *except* for managed fields -- if that is different, then there is no need to compare or modify managed fields.

I believe we must already be doing a deep equal somewhere before writing; if we can at this point somehow pass along a "yes this is really modified" flag with the object, we could skip that later check, making this more neutral.

Reply:
> Has generation been set at this point? If so, you can short-circuit a spec check by looking at that.

Generation happens in PrepareForUpdate, after this point

> I believe we must already be doing a deep equal somewhere before writing; if we can at this point somehow pass along a "yes this is really modified" flag with the object, we could skip that later check, making this more neutral.

The only equality check I know if in the pipeline is the byte-comparison where etcd storage is considering whether the bytes we're about to write are equal. That's too late for this logic (and would include the managedField differences we're trying to avoid here)

Reply:
I still think pre-checking managed fields for changes that are only timestamps and nothing else before doing the full check may have value.

Since this is only triggered by a small number of cases (nonstandard resource.Quantity, what else?), we could restrict this to a hardcoded list of endpoints that could possibly trigger this. (e.g. probably no status update needs to go through this code.)

If we REALLY want to go down into the weeds, that byte comparison could possibly be made a proto-tag-aware comparison and notice when the only thing different is this particular field (even easier for json). We can depend on well-formatted data at that point so potentially that would be very cheap.

The other alternative is just to debug the current code and figure out why allocations are so high, of course.